tool/src/atsp.dart

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// See: Asymmetric Traveling Salesman Problem.
+
+// Strategy for finding optimal overlapping of chunks of a larger table,
+// to save space.
+// Does so by solving traveling salesman/hamiltonian cycle in a graph
+// where the distance between chunks is how little they overlap
+// (chunk length minus overlap size).
+
+/// An asymmetric weighted complete graph.
+///
+/// The vertices are identified by numbers 0 through [vertices] - 1.
+/// Edges are pairs of vertices.
+class Graph {
+  /// Number of vertices.
+  final int vertices;
+
+  /// Table of weights, a list of length `vertices`*`vertices`.
+  final List<int> _table;
+
+  /// Creates a new complete graph with [vertices] vertices.
+  ///
+  /// The initial weights on all edges are [initialWeight].
+  Graph(this.vertices, [int initialWeight = 0])
+      : _table = List<int>.filled(vertices * vertices, initialWeight);
+
+  /// Update the weight on the edges from [fromVertex] to [toVertex].
+  void setWeight(int fromVertex, int toVertex, int newWeight) {
+    _table[fromVertex * vertices + toVertex] = newWeight;
+  }
+
+  /// The weight of the edge from [fromVertex] to [toVertex].
+  int weight(int fromVertex, int toVertex) =>
+      _table[fromVertex * vertices + toVertex];
+
+  /// The cummulative weight of the (sub-)path from `path[from]` to `path[to]`.
+  ///
+  /// If [to] is less than [from], the sub-path is traversed in reverse.
+  /// The values in `path` should be vertices in this graph.
+  int pathWeight(List<int> path, int from, int to) {
+    int weight = 0;
+    int cursor = path[from];
+    int step = from <= to ? 1 : -1;
+    for (int i = from; i != to;) {
+      i += step;
+      int next = path[i];
+      weight += this.weight(cursor, next);
+      cursor = next;
+    }
+    return weight;
+  }
+
+  int get maxWeight => _table.reduce((a, b) => a >= b ? a : b);
+}
+
+/// Optimize a cycle of a graph to minimize the edge weight.
+///
+/// The [cycle] must have the same node as first and last element.
+///
+/// This is an implementation of step of 3-opt, a simple algorithm to
+/// approximate an asymmetric traveling salesman problem (ATSP).
+/// It splits the cycle into three parts and then find the best recombination
+/// of the parts, each potentially reversed.
+bool opt3(Graph graph, List<int> cycle) {
+  // Perhaps optimize the weight computations by creating
+  // a single array of cummulative weights, so any range can be computed
+  // as a difference between two points in that array.
+  for (int i = 1; i < cycle.length; i++) {
+    int wA = cycle[i - 1];
```

Comment:
What is the `w` here short for?

I understand it is "weight" in `wAB`, but I would have expected a word like `vertexA` for this variable.

Reply:
Changed to `nodeA`, `wAB` and `pBC` to distinguish single-step weights from path weights, and nodes from weights.